### PR TITLE
[#WA-2.1]-survey,[#WA-2.7]-survey statistics

### PR DESCRIPTION
**Pull request by maka998**

### CODE REVIEW

**Comment by maka998** [comment]
[(5) WA-2.1 As a patient who was examined I want to be able to rate overall clinic activity, so I can help clinic to improve.](https://trello.com/c/dqqBUyxd/11-5-wa-21-as-a-patient-who-was-examined-i-want-to-be-able-to-rate-overall-clinic-activity-so-i-can-help-clinic-to-improve)

**Comment by maka998** [comment]
[(5) WA-2.7 As an administrator I want to have insight in survey results so I can see patients opinion and improve clinic work.](https://trello.com/c/N3XEXuhA/17-5-wa-27-as-an-administrator-i-want-to-have-insight-in-survey-results-so-i-can-see-patients-opinion-and-improve-clinic-work)

File: WebApplication/Backend/Repositorys/FeedbackRepository.cs [hunk 1]
```diff
@@ -16,7 +16,8 @@ public FeedbackRepository()
         {
             try
             {
-                connection = new MySqlConnection("server=localhost;port=3306;database=mydb;user=root;password=root");
+                connection = new MySqlConnection("server=localhost;database=newdb;user=root;password=root");
+                connection.Open();
```
Comment (milicasiriski) [inline WebApplication/Backend/Repositorys/FeedbackRepository.cs]: I think only one is enough hahaha

Comment (milicasiriski) [reply WebApplication/Backend/Repositorys/FeedbackRepository.cs]: And I don't think it should always be open? But only when you need to make a request? So in the method that gets executes queries

File: WebApplication/Backend/Repositorys/SurveyRepository.cs [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+
+            List<StatisticAuxilaryClass> statistics = new List<StatisticAuxilaryClass>();
+            for (int i = 0; i < 5; i++)
+            {
+                statistics.Add(new StatisticAuxilaryClass());
+            }
+
+            foreach (Survey s in reports)
+            {
+                statistics[0].AverageRating += Double.Parse(s.Question5);
```
Comment (milicasiriski) [inline WebApplication/Backend/Repositorys/SurveyRepository.cs:356]: I have no idea what is going on here but I think you can extract something to a method

File: WebApplication/Backend/Repositorys/SurveyRepository.cs [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+                    statistics[i].AverageRating = statistics[i].AverageRating / reports.Count/5;
+                else if (i == 1)
+                    statistics[i].AverageRating = statistics[i].AverageRating / reports.Count/4;
+                else if (i == 2)
+                    statistics[i].AverageRating = statistics[i].AverageRating / reports.Count/5;
+                else if (i == 3)
+                    statistics[i].AverageRating = statistics[i].AverageRating / reports.Count/2;
+                else if (i == 4)
+                    statistics[i].AverageRating = statistics[i].AverageRating / reports.Count/3;
+            }
```
Comment (milicasiriski) [inline WebApplication/Backend/Repositorys/SurveyRepository.cs]: Why is this in a for loop if the numbers are fixed? Coundn't it just be:
 `statistics[0].AverageRating = statistics[0].AverageRating / reports.Count/5; `
and so on for each field?

File: WebApplication/Backend/Repositorys/SurveyRepository.cs [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+
+
+        ////Aleksa Repovic RA52/2017
+        /// <summary>
+        ///Helping class which rounds double values to 2 decimals
+        ///</summary>
+        ///<returns>
+        ///list of objects containing information about statistics with values rounded to 2 decimals
+        ///</returns>
+        public List<StatisticAuxilaryClass> round2Decimals(List<StatisticAuxilaryClass> input) {
```
Comment (milicasiriski) [inline WebApplication/Backend/Repositorys/SurveyRepository.cs:416]: 💯

File: WebApplication/Backend/Services/SurveyService.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+using System.Threading.Tasks;
+using WebApplication.Backend.Repositorys;
+using WebApplication.Backend.Util;
+
+namespace WebApplication.Backend.Services
+{
+    public class SurveyService
+    {
+        private SurveyRepository surveyRepository;
+        private ISurveyRepository isurveyRepository = new SurveyRepository();
```
Comment (milicasiriski) [inline WebApplication/Backend/Services/SurveyService.cs]: Same as in other 2 PRs, only one field of type ISurveyRepository named surveyRepository

File: WebApplicationTests/UnitTest1.cs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+using System;
+using System.Collections.Generic;
+using WebApplication.Backend.Repositorys;
+using WebApplication.Backend.Services;
+using WebApplication.Backend.Util;
+using Xunit;
+
+namespace WebApplicationTests
+{
+    public class UnitTest1
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: This could be SurveyAndStatisticsServiceTest

File: WebApplicationTests/UnitTest1.cs [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                var stubRepository = new Mock<ISurveyRepository>();
+                var result = true;
+                Survey surveyTest = new Survey
+                ("001", "Bozo Bozic", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3");
+                stubRepository.Setup(m => m.AddNewSurvey(surveyTest)).Returns(result);
+                SurveyService service = new SurveyService(stubRepository.Object);
+                bool result1 = service.AddNewSurvey(surveyTest);
+                Assert.True(result1);
+            }
+        /*
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: delete if it's unused

File: WebApplicationTests/UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+{
+    public class UnitTest1
+    {
+            [Fact]
+            public void AddNewSurveyTest()
+            {
+                var stubRepository = new Mock<ISurveyRepository>();
+                var result = true;
+                Survey surveyTest = new Survey
+                ("001", "Bozo Bozic", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3", "3");
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: You can move test objects for stubs to a field or into a constants class to reuse in all tests

Comment (milicasiriski) [reply WebApplicationTests/UnitTest1.cs]: Separate assign, act and assert parts with an empty line 😄

File: WebApplicationTests/UnitTest1.cs [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+using WebApplication.Backend.Services;
+using WebApplication.Backend.Util;
+using Xunit;
+
+namespace WebApplicationTests
+{
+    public class UnitTest1
+    {
+            [Fact]
+            public void AddNewSurveyTest()
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: test name convention, others have used the sentence notations so something like Adds_new_survey would be appropriate

File: WebApplicationTests/UnitTest1.cs [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+          
+            }
+            */
+            [Fact]
+            public void getStatisticEachQuestion_correct()
+            {
+                using (var mock = AutoMock.GetLoose()) {
+                    mock.Mock<ISurveyRepository>()
+                        .Setup(x => x.getStatisticsEachQuestion())
+                        .Returns(getSampleStatisticsEachQuestion());
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: Nice that it's in the separate method 🥇

File: WebApplicationTests/UnitTest1.cs [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+                return false;
+            else if (p1.Fours != p2.Fours)
+                return false;
+            else if (p1.Fives != p2.Fives)
+                return false;
+   
+            return true;
+        }
+
+            private List<StatisticAuxilaryClass> getSampleStatisticsEachQuestion()
```
Comment (milicasiriski) [inline WebApplicationTests/UnitTest1.cs]: Formatting is kind of weird here, use [this](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-al-formatter?fbclid=IwAR1AiWwyO-qe7JnASpmqKUfwYaGk0vXDW4QsqqgYW9nqC68ULmIuxZf2i8Q#:~:text=In%20an%20existing%20project%2C%20open,shortcut%20Alt%2BShift%2BF) and it should fix it

File: project-generated-code-backend/Backend/Controller/PhysitianControllers/ExamController.cs [hunk 12]
```diff
@@ -35,12 +35,12 @@ public void SaveReport()
         {
             reportService.NewReport(currentReport);
         }
-
+        /*
```
Comment (milicasiriski) [inline project-generated-code-backend/Backend/Controller/PhysitianControllers/ExamController.cs]: Why is this in a comment?

**Review by milicasiriski — CHANGES_REQUESTED** [review]
Could have been 2 PRs 😄 Most mistakes have been repeated by all your team members. Tests are ok just use the naming convention

File: project-generated-code-backend/Backend/Controller/PhysitianControllers/ExamController.cs (hunk 12 above)
Comment (repovic48) [reply project-generated-code-backend/Backend/Controller/PhysitianControllers/ExamController.cs]: This code is not ours.

Comment (milicasiriski) [reply project-generated-code-backend/Backend/Controller/PhysitianControllers/ExamController.cs]: But you put it in a comment

**Review by milicasiriski — APPROVED** [review]

**Review by lukaDoric — APPROVED** [review]

**Comment by maka998** [comment]
[(5) WA-2.1 As a patient who was examined I want to be able to rate overall clinic activity, so I can help clinic to improve.](https://trello.com/c/dqqBUyxd/11-5-wa-21-as-a-patient-who-was-examined-i-want-to-be-able-to-rate-overall-clinic-activity-so-i-can-help-clinic-to-improve)

**Comment by maka998** [comment]
[(5) WA-2.7 As an administrator I want to have insight in survey results so I can see patients opinion and improve clinic work.](https://trello.com/c/N3XEXuhA/17-5-wa-27-as-an-administrator-i-want-to-have-insight-in-survey-results-so-i-can-see-patients-opinion-and-improve-clinic-work)

**Pull request merged**